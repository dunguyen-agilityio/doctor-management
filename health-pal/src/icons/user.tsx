import Svg, { G, Path, SvgProps } from 'react-native-svg'

const User = (props: SvgProps) => {
  return (
    <Svg width="18" height="19" viewBox="0 0 18 19" fill="none" {...props}>
      <G id="vuesax/linear/user">
        <G id="user">
          <Path
            id="Vector"
            d="M9 9.5C11.0711 9.5 12.75 7.82107 12.75 5.75C12.75 3.67893 11.0711 2 9 2C6.92893 2 5.25 3.67893 5.25 5.75C5.25 7.82107 6.92893 9.5 9 9.5Z"
            stroke="#9CA3AF"
            stroke-width="1.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
          <Path
            id="Vector_2"
            d="M15.4426 17C15.4426 14.0975 12.5551 11.75 9.00011 11.75C5.44511 11.75 2.55762 14.0975 2.55762 17"
            stroke="#9CA3AF"
            stroke-width="1.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </G>
      </G>
    </Svg>
  )
}

export default User
