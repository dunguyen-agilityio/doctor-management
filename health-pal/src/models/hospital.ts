import { TImage } from '@/types/image'

import { Doctor } from './doctor'
import { Review } from './review'
import { Strapi } from './strapi'

export class Hospital extends Strapi {
  description!: string
  name!: string
  address!: string
  image?: TImage
  doctors!: Doctor[]
  rating!: number
  reviews!: Review[]
  reviewCounter!: number
  type!: 'Hospital' | 'Clinic'

  constructor({ doctors, reviews, rating, reviewCounter, ...clinic }: Hospital) {
    super(clinic)
    Object.assign(this, clinic)
    this.doctors = doctors ?? []
    this.reviews = reviews ?? []
    this.rating = rating ?? 0
    this.reviewCounter = reviewCounter ?? 0
  }
}

export type THospitalFavorite = Hospital & { favoriteId?: string }
