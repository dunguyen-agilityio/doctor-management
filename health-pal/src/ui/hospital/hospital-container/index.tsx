import { StyleSheet } from 'react-native'

import { Spinner, Stack, View } from 'tamagui'

import { FAVORITE_EMPTY } from '@app/constants'

import { YStack } from '@theme/stack'

import { LoadingIndicator } from '@app/components'
import Empty from '@app/components/empty'
import ErrorState from '@app/components/error'
import useHospitals from '@app/hooks/use-hospitals'
import useMediaQuery from '@app/hooks/use-media-query'
import { FAVORITE_TYPES } from '@app/types/favorite'

import HospitalList from '../hospital-list'

const ItemSeparatorComponent = () => <Stack height={12} />

const HospitalContainer = ({ query }: { query: string }) => {
  const { data, isFetching, isFetchingNextPage, error, refetch, fetchNextPage, hasNextPage } =
    useHospitals(query)

  const { height } = useMediaQuery({ px: 24, height: 256, full: true })

  if ((!data && !isFetching) || error)
    return (
      <ErrorState
        title="Error Loading Favorites"
        message="We couldn't load your favorite Hospital. Please try again."
        onRetry={refetch}
      />
    )

  const ListEmptyComponent = <Empty {...FAVORITE_EMPTY[FAVORITE_TYPES.HOSPITAL]} />

  const ListFooterComponent = hasNextPage ? (
    <YStack paddingVertical={8}>
      <Spinner />
    </YStack>
  ) : null

  return (
    <View flex={1} position="relative">
      {isFetching && !isFetchingNextPage && <LoadingIndicator fullScreen />}
      <HospitalList
        data={data?.data ?? []}
        ItemSeparatorComponent={ItemSeparatorComponent}
        ListEmptyComponent={ListEmptyComponent}
        contentContainerStyle={styles.contentContainerStyle}
        estimatedItemSize={height}
        onEndReached={() => fetchNextPage()}
        ListFooterComponent={ListFooterComponent}
      />
    </View>
  )
}

export default HospitalContainer

const styles = StyleSheet.create({
  contentContainerStyle: {
    paddingHorizontal: 24,
    paddingVertical: 8,
  },
})
