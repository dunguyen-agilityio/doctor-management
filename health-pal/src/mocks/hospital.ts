import { Hospital } from '@app/models/hospital'

export const MOCK_HOSPITALS: Hospital[] = [
  {
    id: 1,
    documentId: '123e4567-e89b-12d3-a456-426614174000',
    name: 'Sunrise Hospital',
    description: 'A premier healthcare facility offering top-tier services in Springfield.',
    address: '123 Main St, Springfield, USA',
    image: { id: 1, url: 'https://picsum.photos/200', alt: 'Sunrise Hospital' },
    rating: 4.2,
    doctors: [],
    reivewCouter: 300,
    reviews: [],
    type: 'Clinic',
  },
  {
    id: 2,
    documentId: '987fcdeb-12ab-34cd-5678-901234567890',
    name: 'Mercy Hospital',
    description: 'Leading medical center in Metropolis.',
    address: '456 Oak Ave, Metropolis, Canada',
    image: { id: 2, url: 'https://picsum.photos/201', alt: 'Mercy Hospital' },
    doctors: [],
    reivewCouter: 300,
    reviews: [],
    rating: 1,
    type: 'Hospital',
  },
]
