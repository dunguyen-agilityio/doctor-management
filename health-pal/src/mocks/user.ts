import { User } from '@/models/user'

export const MOCK_USER: User = {
  id: 1,
  name: 'John Doe',
  nickname: 'jdoe',
  email: 'jdoe@example.com',
  documentId: '123e4567-e89b-12d3-a456-426614174000',
  dateOfBirth: new Date('1990-01-01'),
  gender: true,
  avatar: {
    id: 1,
    url: 'https://example.com/avatar.jpg',
  },
  username: 'jdoe',
  bio: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
}
