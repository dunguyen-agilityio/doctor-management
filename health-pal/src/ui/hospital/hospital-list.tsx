import React, { use } from 'react'
import { FlatListProps } from 'react-native'

import { FlatList } from 'react-native-gesture-handler'

import { ClinicsContext } from '@app/contexts/clinic'
import { Clinic } from '@app/models/clinic'

interface HospitalProps extends Omit<FlatListProps<Clinic>, 'data'> {}

const keyExtractor = (item: Clinic) => item.documentId

const HospitalList = ({ ...otherProps }: HospitalProps) => {
  const clinicContext = use(ClinicsContext)

  return (
    <FlatList
      {...otherProps}
      data={clinicContext}
      keyExtractor={keyExtractor}
      showsHorizontalScrollIndicator={false}
      showsVerticalScrollIndicator={false}
      contentContainerStyle={{ paddingBottom: 16 }}
    />
  )
}

export default HospitalList
