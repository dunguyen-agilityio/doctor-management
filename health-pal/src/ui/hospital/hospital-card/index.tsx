import { useRef } from 'react'
import { GestureResponderEvent } from 'react-native'

import { Card, Separator, ViewProps, XStack } from 'tamagui'

import { useAddFavorite } from '@app/hooks/use-add-favorite'
import { useRemoveFavorite } from '@app/hooks/use-remove-favorite'

import { CloudinaryImage, Heading, Stars, Text } from '@app/components'

import FavoriteButton from '@app/ui/favorite/favorite-button'
import RemoveFavoriteModal from '@app/ui/favorite/remove-favorite-confirm-modal'

import { Hospital as HospitalIcon, Routing } from '@icons'
import LocationOutline from '@icons/location-outline'

import { FAVORITE_TYPES } from '@app/types/favorite'
import { ModalRef } from '@app/types/modal'

import { Hospital } from '@app/models/hospital'
import { useFavoritesStore } from '@app/stores/favorite'

type HospitalExtend = {
  marginLeft?: number
  marginRight?: number
  marginBottom?: number
  width?: ViewProps['width']
}

type HospitalCardProps = Hospital & HospitalExtend

const HospitalCard = ({
  image = require('@/assets/images/banner01.webp'),
  address,
  name,
  rating = 4.2,
  reviewCounter = 23,
  type = 'Hospital',
  id,
  width = 'auto',
  marginBottom,
  marginLeft,
  marginRight,
  ...props
}: HospitalCardProps) => {
  const favoriteId = useFavoritesStore((state) => state.favoriteHospitals[id])

  const { mutate: removeFavorite, isPending: removeFavPending } = useRemoveFavorite(
    FAVORITE_TYPES.HOSPITAL,
    name,
  )

  const { mutate: addFavorite, isPending: addFavPending } = useAddFavorite(
    FAVORITE_TYPES.HOSPITAL,
    name,
  )

  const confirmRef = useRef<ModalRef>(null)

  const handleFavorite = (e: GestureResponderEvent) => {
    e.preventDefault()

    if (favoriteId) {
      confirmRef.current?.open()
      return
    }

    addFavorite(id)
  }

  const handleRemove = () => {
    confirmRef.current?.close()
    removeFavorite(favoriteId)
  }

  const renderCard = (extend: HospitalExtend) => (
    <Card
      elevate
      bordered
      borderRadius="$2"
      overflow="hidden"
      elevation={3}
      shadowColor="$black"
      disabled={!addFavPending || removeFavPending}
      disabledStyle={{ opacity: 0.8 }}
      width={width}
      shadowOffset={{ width: 4, height: 4 }}
      {...props}
      {...extend}>
      <Card.Header padding={0}>
        <CloudinaryImage
          source={{ uri: image?.url }}
          style={{ height: 120 }}
          tabIndex={0}
          aria-label={`Image of ${name}`}
          role="img"
        />
        <FavoriteButton
          favoriteId={favoriteId}
          itemId={id}
          itemName={name}
          type={FAVORITE_TYPES.HOSPITAL}
          top={6}
          right={6}
          position="absolute"
          testID="favorite-button"
          onPress={handleFavorite}
          tabIndex={0}
          aria-label={favoriteId ? `Remove ${name} from favorites` : `Add ${name} to favorites`}
          accessibilityHint={
            favoriteId
              ? 'Removes this hospital from your favorites'
              : 'Adds this hospital to your favorites'
          }
          role="button"
        />
      </Card.Header>

      <Card.Footer padding={12} paddingTop={8} gap="$sm" flexDirection="column" tabIndex={0}>
        <Heading size="small" numberOfLines={1}>
          {name}
        </Heading>

        <XStack alignItems="center" gap="$sm">
          <LocationOutline testID="location-icon" />
          <Text size="extraSmall" numberOfLines={1}>
            {address}
          </Text>
        </XStack>

        <XStack alignItems="center" gap="$sm">
          <Stars stars={rating} max={1} />
          <Text size="extraSmall">{`(${reviewCounter} Reviews)`}</Text>
        </XStack>

        <Separator backgroundColor="$grey200" />

        <XStack justifyContent="space-between">
          <XStack alignItems="center" gap="$sm">
            <Routing testID="routing-icon" />
            <Text size="extraSmall">2.5 km/40min</Text>
          </XStack>

          <XStack alignItems="center" gap="$sm">
            <HospitalIcon testID="hospital-icon" />
            <Text size="extraSmall">{type}</Text>
          </XStack>
        </XStack>
      </Card.Footer>
    </Card>
  )

  const content = (
    <>
      {renderCard({ width, marginBottom, marginLeft, marginRight })}
      <RemoveFavoriteModal onConfirm={handleRemove} ref={confirmRef}>
        {renderCard({ width: 'auto' })}
      </RemoveFavoriteModal>
    </>
  )

  return content
}

export default HospitalCard
