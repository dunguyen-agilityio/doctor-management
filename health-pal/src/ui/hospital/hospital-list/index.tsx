import { Hospital } from '@/models/hospital'
import { keyExtractor } from '@/utils/list'
import { FlashList, FlashListProps } from '@shopify/flash-list'

import HospitalCard from '../hospital-card'

export interface HospitalProps extends Omit<FlashListProps<Hospital>, 'renderItem'> {
  renderItem?: FlashListProps<Hospital>['renderItem']
}

const HospitalList = ({ ...otherProps }: HospitalProps) => {
  const renderItem = ({ item }: { item: Hospital }) => <HospitalCard {...item} />

  return (
    <FlashList
      renderItem={renderItem}
      keyExtractor={keyExtractor}
      showsHorizontalScrollIndicator={false}
      showsVerticalScrollIndicator={false}
      testID="hospital-list"
      {...otherProps}
    />
  )
}

export default HospitalList
