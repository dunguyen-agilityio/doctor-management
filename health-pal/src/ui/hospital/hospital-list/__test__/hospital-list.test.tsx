import { MOCK_HOSPITALS } from '@/mocks/hospital'
import { MOCK_USER } from '@/mocks/user'
import { render, screen } from '@utils-test'

import HospitalList from '..'

jest.mock('@/hooks/use-require-auth', () => ({
  useRequireAuth: jest.fn().mockReturnValue({
    session: { user: MOCK_USER },
  }),
}))

const [name1, name2] = MOCK_HOSPITALS.map((item) => item.name)

describe('HospitalList', () => {
  it('renders hospital items', () => {
    render(<HospitalList data={MOCK_HOSPITALS} />)

    expect(screen.getByText(name1)).toBeTruthy()
    expect(screen.getByText(name2)).toBeTruthy()
  })

  it('renders nothing if data is empty', () => {
    const { toJSON } = render(<HospitalList data={[]} />)
    expect(toJSON()).toMatchSnapshot()
  })
})
