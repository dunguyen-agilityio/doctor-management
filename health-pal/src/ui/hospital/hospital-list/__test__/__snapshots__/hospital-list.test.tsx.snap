// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HospitalList renders nothing if data is empty 1`] = `
[
  <View
    aria-label="Notifications (F8)"
    pointerEvents="box-none"
    role="region"
    style={
      {
        "bottom": 0,
        "flexDirection": "column",
        "left": 0,
        "maxWidth": "100%",
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 100000,
      }
    }
    tabIndex={-1}
  >
    <View
      focusable={false}
      pointerEvents="box-none"
      style={
        {
          "flexDirection": "column",
          "maxWidth": "100%",
          "position": "absolute",
        }
      }
    />
  </View>,
  <View
    collapsable={false}
    onLayout={[Function]}
    style={
      {
        "flex": 1,
        "overflow": "hidden",
      }
    }
  >
    <RCTScrollView
      collapsable={false}
      keyExtractor={[Function]}
      maintainVisibleContentPosition={
        {
          "minIndexForVisible": 0,
        }
      }
      onScroll={[Function]}
      scrollEventThrottle={0.0001}
      showsHorizontalScrollIndicator={false}
      showsVerticalScrollIndicator={false}
      testID="hospital-list"
    >
      <View>
        <View
          style={
            {
              "height": 0,
              "left": 0,
              "position": "absolute",
              "top": 1000000,
            }
          }
        />
        <View
          style={
            {
              "height": 0,
              "minHeight": undefined,
              "width": undefined,
            }
          }
        />
        <View
          style={false}
        />
      </View>
    </RCTScrollView>
  </View>,
]
`;
